File: src/main/java/com/zerozero/store/infrastructure/kakao/search/application/KakaoStoreSearcher.java
package com.zerozero.store.infrastructure.kakao.search.application;

import com.zerozero.store.domain.response.StoreSearchResponse;
import com.zerozero.store.domain.service.StoreSearcher;
import com.zerozero.store.exception.StoreErrorType;
import com.zerozero.store.exception.StoreException;
import com.zerozero.store.infrastructure.kakao.search.request.KakaoSearchRequest;
import com.zerozero.store.infrastructure.kakao.search.response.KakaoSearchResponseWrapper;
import lombok.RequiredArgsConstructor;
import lombok.extern.log4j.Log4j2;
import org.springframework.stereotype.Component;

import java.util.Arrays;
import java.util.List;
import java.util.stream.Collectors;

@Component
@RequiredArgsConstructor
@Log4j2
public class KakaoStoreSearcher implements StoreSearcher {

    private final KakaoSearchClient kakaoSearchClient;
    private static final Integer DEFAULT_RADIUS = 2000;

    @Override
    public List<StoreSearchResponse> search(String query) {
        KakaoSearchRequest request = KakaoSearchRequest.builder()
                .query(query)
                .build();
        return executeSearch(request);
    }

    @Override
    public List<StoreSearchResponse> searchByLocation(String query, double longitude, double latitude) {
        KakaoSearchRequest request = KakaoSearchRequest.builder()
                .query(query)
                .x(String.valueOf(longitude))
                .y(String.valueOf(latitude))
                .radius(DEFAULT_RADIUS)
                .build();
        return executeSearch(request);
    }

    private List<StoreSearchResponse> executeSearch(KakaoSearchRequest request) {
        KakaoSearchResponseWrapper kakaoSearchResponseWrapper = kakaoSearchClient.search(request);

        if (kakaoSearchResponseWrapper == null ||
                kakaoSearchResponseWrapper.getKakaoSearchMeta() == null ||
                kakaoSearchResponseWrapper.getKakaoSearchMeta().getTotalCount() == 0) {
            log.error("[KakaoStoreSearcher] Kakao search response is null or empty");
            throw new StoreException(StoreErrorType.NOT_EXIST_SEARCH_RESPONSE);
        }

        return Arrays.stream(kakaoSearchResponseWrapper.getKakaoSearchResponses())
                .map(StoreSearchResponse::from)
                .collect(Collectors.toList());
    }
}

File: src/main/java/com/zerozero/store/infrastructure/kakao/search/application/KakaoStoreSearcher.kt
package com.zerozero.store.infrastructure.kakao.search.application

import com.zerozero.store.domain.response.StoreSearchResponse
import com.zerozero.store.domain.service.StoreSearcher
import com.zerozero.store.exception.StoreErrorType
import com.zerozero.store.exception.StoreException
import com.zerozero.store.infrastructure.kakao.search.request.KakaoSearchRequest
import com.zerozero.store.infrastructure.kakao.search.response.KakaoSearchResponseWrapper
import lombok.RequiredArgsConstructor
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Component

@Component
@RequiredArgsConstructor
class KakaoStoreSearcher(
    private val kakaoSearchClient: KakaoSearchClient,
) : StoreSearcher {

    private val log: Logger = LoggerFactory.getLogger(javaClass)

    override fun search(query: String): List<StoreSearchResponse> {
        val request = KakaoSearchRequest(
            query = query
        )
        return executeSearch(request)
    }

    override fun searchByLocation(query: String, longitude: Double, latitude: Double): List<StoreSearchResponse> {
        val request = KakaoSearchRequest(
            query = query,
            x = longitude.toString(),
            y = latitude.toString(),
            radius = DEFAULT_RADIUS
        )
        return executeSearch(request)
    }

    private fun executeSearch(request: KakaoSearchRequest): List<StoreSearchResponse> {
        val response: KakaoSearchResponseWrapper = kakaoSearchClient.search(request)

        if (response.kakaoSearchMeta.totalCount == 0) {
            log.error("[KakaoStoreSearcher] Kakao search response is null or empty")
            throw StoreException(StoreErrorType.NOT_EXIST_SEARCH_RESPONSE)
        }

        return response.kakaoSearchResponses.map(StoreSearchResponse::from)
    }

    companion object {
        private const val DEFAULT_RADIUS = 2000
    }
}
